Frends.Edifact.CreateFromJson/Frends.Edifact.CreateFromJson/Edifact.cs

```diff
@@ -0,0 +1,184 @@
+            WriteEdiMessages(ediXml, documentType, writer);
+
+            // UNZ handling - we could not find a directly write UNZ with Edifabric.
+            // The INTERCHANGE CONTROL COUNT will just default to 1 and
+            // INTERCHANGE CONTROL REFERENCE will be taken from UNB.
+        }
+
+        return ReadStreamToEnd(stream);
+    }
+
+    private static string ReadStreamToEnd(MemoryStream stream)
+    {
+        stream.Position = 0;
+        using var reader = new StreamReader(stream, Encoding.Default);
+        return reader.ReadToEnd();
+    }
+
+    private static Type TypeFactory(XElement unh)
+    {
+        if (unh == null)
+        {
+            throw new ArgumentException("EDIFACT message does not contain UNH header");
+        }
+
+        var unhHeader = DeserializePart<UNH>(unh);
+        var edifactVersion =
+            $"{unhHeader.MessageIdentifier_02.MessageVersionNumber_02}" +
+            $"{unhHeader.MessageIdentifier_02.MessageReleaseNumber_03}";
+        var typeName = $"TS{unhHeader.MessageIdentifier_02.MessageType_01}";
+        var assembly = LoadEdifactVersion(edifactVersion);
+        var returnType = FindEdifactMessageType(edifactVersion, typeName, assembly);
+        return returnType;
+    }
+
+    private static Type FindEdifactMessageType(string edifactVersion, string typeName, Assembly assembly)
+    {
+        var returnType = assembly.ExportedTypes.FirstOrDefault(x => x.Name == typeName);
+        if (returnType == null)
+        {
+            throw new ArgumentException(
+                $"Edifact message type {typeName} was not found in " +
+                $"Edifact version {edifactVersion}");
+        }
+
+        return returnType;
+    }
+
+    private static Assembly LoadEdifactVersion(string edifactVersion)
+    {
+        var assemblyName = $"Frends.Edifabric.Templates.Edifact.{edifactVersion}";
+        try
+        {
+            return Assembly.Load(assemblyName);
+        }
+        catch (Exception ex)
+        {
+            throw new ArgumentOutOfRangeException(
+                $"Version {edifactVersion} is not supported. " +
+                $"See inner exception for details.", ex);
+        }
+    }
+
+    private static Type DetermineDocumentType(XElement ediXml)
+    {
+        var partUNH = ediXml.DescendantsAndSelf("UNH").First();
+        var documentType = TypeFactory(partUNH);
+        return documentType;
+    }
+
+    private static IEdiItem? DeserializeXElement(Type type, XElement xElement)
+    {
+        var serializer = new XmlSerializer(type);
+        using var reader = xElement.CreateReader();
+        var deserialized = (IEdiItem?)serializer.Deserialize(reader);
+        return deserialized;
+    }
+
+    private static void WriteEdiMessages(XElement ediXml, Type documentType, EdifactWriter writer)
+    {
+        var ediMessages = ediXml.DescendantsAndSelf(documentType.Name.ToString());
+        foreach (var xElement in ediMessages)
```

Comment:
Since here is an iteration I would advice passing cancellation token and adding a cancellation check here to enable termination of a process.